From 826324ba7fc03911ca8987532994a9663ba219ac Mon Sep 17 00:00:00 2001
From: 20Finger-Squared <orhylie@gmail.com>
Date: Fri, 2 Jan 2026 23:17:33 +0000
Subject: [PATCH] keymodes

---
 dwm.c | 104 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 104 insertions(+)

diff --git a/dwm.c b/dwm.c
index 4f345ee..4bfe0c4 100644
--- a/dwm.c
+++ b/dwm.c
@@ -59,6 +59,7 @@
 /* enums */
 enum { CurNormal, CurResize, CurMove, CurLast }; /* cursor */
 enum { SchemeNorm, SchemeSel }; /* color schemes */
+enum { ModeCommand, ModeInsert };
 enum { NetSupported, NetWMName, NetWMState, NetWMCheck,
        NetWMFullscreen, NetActiveWindow, NetWMWindowType,
        NetWMWindowTypeDialog, NetClientList, NetLast }; /* EWMH atoms */
@@ -98,6 +99,13 @@ struct Client {
 	Window win;
 };
 
+typedef struct {
+	unsigned int mod[4];
+	KeySym keysym[4];
+	void (*func)(const Arg *);
+	const Arg arg;
+} Command;
+
 typedef struct {
 	unsigned int mod;
 	KeySym keysym;
@@ -151,6 +159,7 @@ static void buttonpress(XEvent *e);
 static void checkotherwm(void);
 static void cleanup(void);
 static void cleanupmon(Monitor *mon);
+static void clearcmd(const Arg *arg);
 static void clientmessage(XEvent *e);
 static void configure(Client *c);
 static void configurenotify(XEvent *e);
@@ -176,6 +185,7 @@ static void grabbuttons(Client *c, int focused);
 static void grabkeys(void);
 static void incnmaster(const Arg *arg);
 static void keypress(XEvent *e);
+static void keypresscmd(XEvent *e);
 static void killclient(const Arg *arg);
 static void manage(Window w, XWindowAttributes *wa);
 static void mappingnotify(XEvent *e);
@@ -199,6 +209,8 @@ static void sendmon(Client *c, Monitor *m);
 static void setclientstate(Client *c, long state);
 static void setfocus(Client *c);
 static void setfullscreen(Client *c, int fullscreen);
+static void setinsertmode(void);
+static void setkeymode(const Arg *arg);
 static void setlayout(const Arg *arg);
 static void setmfact(const Arg *arg);
 static void setup(void);
@@ -241,6 +253,8 @@ static int sw, sh;           /* X display screen geometry width, height */
 static int bh;               /* bar height */
 static int lrpad;            /* sum of left and right padding for text */
 static int (*xerrorxlib)(Display *, XErrorEvent *);
+static unsigned int cmdmod[4];
+static unsigned int keymode = ModeCommand;
 static unsigned int numlockmask = 0;
 static void (*handler[LASTEvent]) (XEvent *) = {
 	[ButtonPress] = buttonpress,
@@ -264,6 +278,7 @@ static Cur *cursor[CurLast];
 static Clr **scheme;
 static Display *dpy;
 static Drw *drw;
+static KeySym cmdkeysym[4];
 static Monitor *mons, *selmon;
 static Window root, wmcheckwin;
 
@@ -511,6 +526,17 @@ cleanupmon(Monitor *mon)
 	free(mon);
 }
 
+void
+clearcmd(const Arg *arg)
+{
+	unsigned int i;
+
+	for (i = 0; i < LENGTH(cmdkeysym); i++) {
+		cmdkeysym[i] = 0;
+		cmdmod[i] = 0;
+	}
+}
+
 void
 clientmessage(XEvent *e)
 {
@@ -951,6 +977,13 @@ grabbuttons(Client *c, int focused)
 void
 grabkeys(void)
 {
+	if (keymode == ModeCommand) {
+		XUngrabKey(dpy, AnyKey, AnyModifier, root);
+		XGrabKeyboard(dpy, root, True, GrabModeAsync, GrabModeAsync, CurrentTime);
+		return;
+	}
+
+	XUngrabKeyboard(dpy, CurrentTime);
 	updatenumlockmask();
 	{
 		unsigned int i, j, k;
@@ -1002,6 +1035,11 @@ keypress(XEvent *e)
 	KeySym keysym;
 	XKeyEvent *ev;
 
+	if (keymode == ModeCommand) {
+		keypresscmd(e);
+		return;
+	}
+
 	ev = &e->xkey;
 	keysym = XKeycodeToKeysym(dpy, (KeyCode)ev->keycode, 0);
 	for (i = 0; i < LENGTH(keys); i++)
@@ -1011,6 +1049,53 @@ keypress(XEvent *e)
 			keys[i].func(&(keys[i].arg));
 }
 
+void
+keypresscmd(XEvent *e) {
+	unsigned int i, j;
+	int matches = 0;
+	KeySym keysym;
+	XKeyEvent *ev;
+
+	ev = &e->xkey;
+	keysym = XKeycodeToKeysym(dpy, (KeyCode)ev->keycode, 0);
+	if (XK_Shift_L <= keysym && keysym <= XK_Hyper_R) {
+		return;
+	}
+
+	for (i = 0; i < LENGTH(cmdkeys); i++) {
+		if (keysym == cmdkeys[i].keysym
+		&& CLEANMASK(cmdkeys[i].mod) == CLEANMASK(ev->state)
+		&& cmdkeys[i].func) {
+			cmdkeys[i].func(&(cmdkeys[i].arg));
+			return;
+		}
+	}
+
+	for (j = 0; j < LENGTH(cmdkeysym); j++) {
+		if (cmdkeysym[j] == 0) {
+			cmdkeysym[j] = keysym;
+			cmdmod[j] = ev->state;
+			break;
+		}
+	}
+
+	for (i = 0; i < LENGTH(commands); i++) {
+		matches = 0;
+		for (j = 0; j < LENGTH(cmdkeysym); j++) {
+			if (cmdkeysym[j] == commands[i].keysym[j]
+			&& CLEANMASK(cmdmod[j]) == CLEANMASK(commands[i].mod[j]))
+				matches++;
+		}
+		if (matches == LENGTH(cmdkeysym)) {
+			if (commands[i].func)
+				commands[i].func(&(commands[i].arg));
+			clearcmd(NULL);
+			return;
+		}
+	}
+}
+
+
 void
 killclient(const Arg *arg)
 {
@@ -1441,6 +1526,24 @@ setclientstate(Client *c, long state)
 		PropModeReplace, (unsigned char *)data, 2);
 }
 
+void
+setinsertmode()
+{
+	keymode = ModeInsert;
+	clearcmd(NULL);
+	grabkeys();
+}
+
+void
+setkeymode(const Arg *arg)
+{
+	if(!arg)
+		return;
+	keymode = arg->ui;
+	clearcmd(NULL);
+	grabkeys();
+}
+
 int
 sendevent(Client *c, Atom proto)
 {
@@ -1648,6 +1751,7 @@ spawn(const Arg *arg)
 {
 	struct sigaction sa;
 
+	setinsertmode();
 	if (arg->v == dmenucmd)
 		dmenumon[0] = '0' + selmon->num;
 	if (fork() == 0) {
-- 
2.50.1

